src/mame/drivers/minitel_2_rpic.cpp

```diff
@@ -0,0 +1,475 @@
+	and a EF9345 like semi graphic video chip.
+
+	The current implementation is an Minitel 2 from "La RADIOTECHNIQUE PORTENSEIGNE" / RPIC (Philips)
+	You can found more informations about this hardware there :
+	http://hxc2001.free.fr/minitel
+
+	What is implemented and working :
+
+	- Main MCU
+	- Video output
+	- Keyboard
+
+	What is not yet implemented :
+
+	- Modem and sound output.
+	- The rear serial port.
+	- Parameters I2C 24C02 EEPROM.
+
+	The original firmware and the experimental demo rom are currently both working.
+
+	Please note the current special function keys assignation :
+
+	F1 -> Suite
+	F2 -> Retour
+	F3 -> Envoi
+	F4 -> Repetition
+	F5 -> TEL
+	F6 -> Guide
+	F7 -> Sommaire
+	F8 -> Connexion/Fin
+	F9 -> Fonction
+	F10-> ON / OFF
+
+	With the official ROM you need to press F10 to switch on the CRT.
+
+****************************************************************************/
+
+#include "emu.h"
+
+#include "cpu/mcs51/mcs51.h"
+#include "video/ef9345.h"
+
+#include "screen.h"
+#include "softlist.h"
+
+// IO expander latch usage definitions
+#define CTRL_REG_DTMF  0x02
+#define CTRL_REG_MCBC  0x04
+#define CTRL_REG_OPTO  0x08
+#define CTRL_REG_RELAY 0x10
+#define CTRL_REG_CRTON 0x20
+
+// 80C32 Port IO usage definitions
+#define PORT_1_KBSERIN 0x01
+#define PORT_1_MDM_DCD 0x02
+#define PORT_1_MDM_PRD 0x04
+#define PORT_1_MDM_TXD 0x08
+#define PORT_1_MDM_RTS 0x10
+#define PORT_1_KBLOAD  0x20
+#define PORT_1_SCL     0x40
+#define PORT_1_SDA     0x80
+
+//#define DEBUG_MINITEL 1
+
+class minitel_state : public driver_device
+{
+public:
+	minitel_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_maincpu(*this, "maincpu")
+		, m_ts9347(*this, "ts9347")
+		, m_palette(*this, "palette")
+		{
+		}
+
+	DECLARE_WRITE8_MEMBER(port_w);
+	DECLARE_READ8_MEMBER(port_r);
+
+	DECLARE_WRITE8_MEMBER(dev_crtl_reg_w);
+	DECLARE_READ8_MEMBER(dev_keyb_ser_r);
+
+	DECLARE_READ8_MEMBER ( ts9347_io_r );
+	DECLARE_WRITE8_MEMBER ( ts9347_io_w );
+	required_device<cpu_device> m_maincpu;
+	required_device<ts9347_device> m_ts9347;
+	required_device<palette_device> m_palette;
+
+	virtual void machine_start() override;
+
+	char port0, port1, port2, port3;
+
+	int keyboard_para_ser;
+	unsigned char keyboard_x_row_reg;
+
+	unsigned char last_ctrl_reg;
+
+	TIMER_DEVICE_CALLBACK_MEMBER(minitel_scanline);
+};
+
+void minitel_state::machine_start()
+{
+	m_palette->set_pen_color( 0, 0, 0, 0);
+	m_palette->set_pen_color( 1, 86, 86, 86);
+	m_palette->set_pen_color( 2, 172, 172, 172);
+	m_palette->set_pen_color( 3, 255, 255, 255);
+	m_palette->set_pen_color( 4, 44, 44, 44);
+	m_palette->set_pen_color( 5, 86, 86, 86);
+	m_palette->set_pen_color( 6, 172, 172, 172);
+	m_palette->set_pen_color( 7, 255, 255, 255);
+}
+
+WRITE8_MEMBER(minitel_state::port_w)
+{
+	#ifdef DEBUG_MINITEL
+	printf("port_w: write %02X to PORT (offset=%02X)\n", data, offset);
+	#endif
```

Comment:
Please use the compile-time configurable logging feature from `logmacro.h` for this kind of logging - using `#ifdef` allows the expressions to rot when not enabled since the compiler doesn't even look at them.